NTNET Rework so machines can work in harmony!

WarlockD opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Mostly backend support.  the network component is hardly used so preparing it so it could be used to network and communicate to one another.  Quiet possibly replacing the snowflake radio device used in hardware.

## Why It's Good For The Game

Eventually we will be able to link devices not based on ranged by by ip address!

Still a work in progress but I got a clear idea on how it will  work

## Changelog
:cl:
add: IP addresses and routing start.  Make it a bit more apporable to new players
add: Added Tablet's as a "tool"
del: Removed ntnet services, never used and implementation is kind of stupid
tweak: ntnet is been subclassed so we can have more than one network
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

WarlockD committed:
Ok, it works at THIS point subclassing nttd

WarlockD committed:
Update ntnet_interface.dm

WarlockD committed:
Net rework

zxaber commented:
Limiting to tablets is a bit strange. We have other modular computers (Laptops, and PDAs will be when I get done with everything).

The eventual plan is a method for tablets to be used to scan/interact with items in the world via the attack chain, and then passing that action onto the active program. Can this be altered to that end?

WarlockD commented:
Honestly I wanted to use the multi tool to do the bare bones interface.  Like configure IP address, check if the thing is broken etc. Just people kept suggesting using tablets.  I eventually want a command line interface like set up.  Switching to a multi tool would be simple for me to do so I don't step on other prs.

zxaber commented:
If you want to do a multitool for now, as a temporary measure, I can move it over to modular computers once I do tapping support. We actually already have an NT Net diagnostics tool (Wire Carp) that we could add this functionality to.

WarlockD committed:
Removed the tablet and going ot use mutli tool

XDTM commented:
If the changes are backend-only there's no need for a changelog

Rohesie commented on code/__HELPERS/_logging.dm:
```diff
@@ -151,6 +151,10 @@
 	if (CONFIG_GET(flag/log_telecomms))
 		WRITE_LOG(GLOB.world_telecomms_log, "TCOMMS: [text]")
 
+/proc/log_network(text)
+	if (CONFIG_GET(flag/log_network))
+		WRITE_LOG(GLOB.network_log, "NETWORK: [text]")
```
```suggestion
#DEFINE LOG_NETWORK(text) if (CONFIG_GET(flag/log_network)) { WRITE_LOG(GLOB.network_log, "NETWORK: [text]") }
		
```
Optional change. I know it's not how it's done for the other procs, but there's no reason for these to be procs instead of defines.

Rohesie commented on code/controllers/subsystem/processing/networks.dm:
```diff
@@ -6,46 +6,57 @@ PROCESSING_SUBSYSTEM_DEF(networks)
+		if(!networks_by_id[network])
+			if(!parent)
+				parent = station_network
+			net = new/datum/ntnet/(network, parent) // always a parrent to the main network
```
```suggestion
			net = new /datum/ntnet/(network, parent) // always a parrent to the main network
```
Clarity

Rohesie commented on code/controllers/subsystem/processing/networks.dm:
```diff
@@ -6,46 +6,57 @@ PROCESSING_SUBSYSTEM_DEF(networks)
+	do
+		var/string = md5("[num2text(rand(HID_RESTRICTED_END, 999999999), 12)]")
+		if(!string)
+			log_runtime("Could not generagea m5 hash from address, problem with md5?")
```
```suggestion
			CRASH("Could not generagea m5 hash from address, problem with md5?")
```

Rohesie commented on code/controllers/subsystem/processing/networks.dm:
```diff
@@ -6,46 +6,57 @@ PROCESSING_SUBSYSTEM_DEF(networks)
+	do
+		var/string = md5("[num2text(rand(HID_RESTRICTED_END, 999999999), 12)]")
+		if(!string)
+			log_runtime("Could not generagea m5 hash from address, problem with md5?")
+			return		//errored
```
```suggestion
			CRASH("Could not generagea m5 hash from address, problem with md5?")
```

Rohesie commented on code/controllers/subsystem/processing/networks.dm:
```diff
@@ -6,46 +6,57 @@ PROCESSING_SUBSYSTEM_DEF(networks)
-	if(interfaces_by_id[.])
-		return resolve_collisions? make_address("[num2text(rand(HID_RESTRICTED_END, 999999999), 12)]"):null
+	// lets get rid of recursion
+	var/static/list/collision_check = list()
```
Might be better to have a global managed list, for better debugging.

WarlockD replied:
True, but kind of internaly debated about having a function handle its own checks on if a number has been used or not.

Rohesie commented on code/controllers/subsystem/processing/networks.dm:
```diff
@@ -6,46 +6,57 @@ PROCESSING_SUBSYSTEM_DEF(networks)
+	while(collision_check[.])
+	collision_check[.] = TRUE
+
+
+/datum/controller/subsystem/processing/networks/proc/register_interface(datum/component/ntnet_interface/I, network=null)
+		I.hardware_id = "[get_next_HID()]"
+		interfaces_by_hardware_id[I.hardware_id] = I
+		var/datum/ntnet/net = register_network(network)	// default to station
+		net.interface_connect(I)
+
+/datum/controller/subsystem/processing/networks/proc/unregister_interface(datum/component/ntnet_interface/I)
```
Document the procs, and please use a better variable name than `I`

Rohesie commented on code/datums/components/ntnet_interface.dm:
```diff
@@ -1,4 +1,117 @@
+	return src
+
+/datum/ip/proc/operator&(datum/ip/B)
+	return new/datum/ip(upper & B.upper,lower & B.lower)
```
```suggestion
	return new /datum/ip(upper & B.upper,lower & B.lower)
```
Clarity.

Jared-Fogle commented on code/datums/components/ntnet_interface.dm:
```diff
@@ -1,4 +1,117 @@
+/datum/ip/proc/to_string()
+	return text
+
+/datum/ip/proc/operator&=(datum/ip/B)
+	upper &= B.upper
+	lower &= B.lower
+	return src
+
+/datum/ip/proc/operator&(datum/ip/B)
+	return new/datum/ip(upper & B.upper,lower & B.lower)
+
+/datum/ip/proc/operator~=(datum/ip/B)
+	return upper == B.upper && lower == B.lower
```
Can you make these not use operator overloads? I don't think they particularly help readability, and they definitely make it harder to track down implementations--I can't just alt click a proc name with these.

WarlockD replied:
Sure, I am changing how ips work

WarlockD committed:
more updates, still not sure about ip address

Tharcoonvagh commented:
Amazing

WarlockD commented:
ah crap I forgot I did this

WarlockD committed:
Removed IP, going with kind of an IPX networking

WarlockD committed:
Ok more streamlining, removed dns for now, trying to get this to work before adding features

WarlockD committed:
Well we compile and the doors still work, thats a plus.  Now lets convert atmos!

WarlockD committed:
Simplfied the networks farther.  Going to make this simple then expand latter

WarlockD committed:
Merge branch 'master' into machine_network

WarlockD committed:
Bit more cleanup and simplification.

Alright.  Now to tackle atmos replacement

WarlockD committed:
Alright! atmos has been changed FOREVHER HAHASHAH

WarlockD committed:
so it all compiles...but the network dosn't work

meh

WarlockD committed:
Mother fucking fucker.

Modified the maps so they have the right tags.  Still having runtimes on the network creation, not sure why.  tracing it now

WarlockD committed:
The god damn thing works.

Still have alot of snowflakes to go with

WarlockD committed:
Revert "The god damn thing works."

This reverts commit 0e7809071b2f970a4bf9247edf33b32a735fdcce.

WarlockD committed:
Changed back to id_tag so I don't have to update all those damn maps.

WarlockD committed:
Merge branch 'master' into machine_network

WarlockD commented:
Ok this list of reviewers is getting silly.  Once I get this into a testable state I will repr it...hopefuly

WarlockD committed:
Just more changes to some of the netmethod.

Ugh really? no operator overloads, meh.
Anyway fixing up ports now.  I already see preformance increase with the air alarm when you put down like 70 scrubbers in an area...not that there is a reason to

WarlockD committed:
Merge branch 'machine_network' of https://github.com/WarlockD/tgstation into machine_network

WarlockD committed:
ok

Refactoring datalink, got a better way to do it

WarlockD committed:
Ugh, I cannot belelve I didn't see this...just ugh.

WarlockD committed:
uuuuuugggggg

AnturK commented on .vscode/settings.json:
```diff
@@ -12,5 +12,11 @@
+	"gitlens.advanced.blame.customArguments": [
+		"-w"
+	],
+	"cSpell.words": [
```
Ahem

AnturK commented on _maps/RandomRuins/SpaceRuins/deepstorage.dmm:
```diff
@@ -55,6 +55,7 @@
 /area/ruin/space/has_grav/deepstorage/crusher)
 "ao" = (
 /obj/machinery/atmospherics/components/unary/vent_pump/on{
+	network_id = NETWORK_CHARLIE_ATMOS;
```
Defines in dmm

AnturK commented on code/controllers/subsystem/processing/networks.dm:
```diff
@@ -1,51 +1,135 @@
+	return ..()
+
+/datum/controller/subsystem/processing/networks/fire(resumed = 0)
+	// so life sucks.  Can't be in Initialize because Initialize is run async and we must start
```
What are you talking about, subsystems have strict initialization order.

WarlockD replied:
Doesn't.  atmosinit() runs during mapload, async during the mapping process BEFORE Initialize.  On each template.  So yes, while Initialize does load after that map has loaded it screwed me up because vents and scrubbers soft'ed Initialized before that.  I refractored a lot of atmos and now I am getting rid of that subsystem hack so we just need to use Initialize.  It just took to much digging into how the atmos.dm subsystem worked and pissed me off.

AnturK commented on code/controllers/subsystem/processing/networks.dm:
```diff
@@ -1,51 +1,135 @@
+// verifies the network name has the right letters in it.  Might need to expand unwanted punctuations
+
+/datum/controller/subsystem/processing/networks/proc/verify_network_name(name)
+	return istext(name) && length(name) > 0 && name[1] != "." && name[length(name)] != "."&& findtext(name,"\[ \n-;:\]*") == 0
```
If it's alphanumerics in some specific format only do whitelist with regex instead of this.

AnturK reviewed (commented):
I'm assuming this is not done yet ?

WarlockD committed:
airalarm works now ughh.

Finaly got ports working on the air alarm.  Need to refactor the air alarm tgui interface but going to hold off for another pr.  Network first, revamp latter

WarlockD committed:
ok added station name.  Still having issues with control

WarlockD commented:
Yess this is still WIP.  But I think I am getting it finialized now and finaly fixing the init order.

WarlockD committed:
We are DONE...mostly

Ok, now the nasty part of moving ALL this code over to a fresh pr and branch...meh

WarlockD committed:
more changes more hell

LemonInTheDark commented:
How's she doing?

Cyberboss commented:
Finish this pls

WarlockD commented:
Ok closing this pr.  I touched WAY to many files for what this does.  Re pring it with a smaller scope then expand it to individual machine types latter.
